### PR TITLE
compiler: optimize tests

### PR DESCRIPTION
**Pull request by fyrchik**
The longest compiler tests are the ones that check some specific construction using it in multiple ways.
We can put such tests in functions in a single file.
For other tests this is more tricky because some of them are specifically crafted to handle particular import/global variables
configuration, so leave them as they are.

```
# master
ok      github.com/nspcc-dev/neo-go/pkg/compiler        15.357s
# PR
ok      github.com/nspcc-dev/neo-go/pkg/compiler        9.553s
```

Related #2379 .

Signed-off-by: Evgeniy Stratonikov <evgeniy@nspcc.ru>

### CODE REVIEW

**Review by roman-khimov — APPROVED** [review]

File: pkg/compiler/for_test.go [hunk 1]
```diff
@@ -732,43 +681,47 @@ func TestForLoopRangeMap(t *testing.T) {
...
+	srcBuilder := bytes.NewBuffer([]byte("package testcase\n"))
+	for i, tc := range binaryExprTestCases {
+		srcBuilder.WriteString(fmt.Sprintf(tc.src, i))
+	}
+
+	ne, di, err := compiler.CompileWithOptions("file.go", strings.NewReader(srcBuilder.String()), nil)
+	require.NoError(t, err)
 
-	tests := []struct {
+	for i, tc := range testCases {
```
Comment (roman-khimov) [inline pkg/compiler/for_test.go:712]: `forLoopTestCases`

File: pkg/compiler/for_test.go [hunk 2]
```diff
@@ -732,43 +681,47 @@ func TestForLoopRangeMap(t *testing.T) {
...
-			i++
-			j++
-			%s
-		}
-		return i
-	}`
+func TestForLoop(t *testing.T) {
+	var testCases []testCase
+	srcBuilder := bytes.NewBuffer([]byte("package testcase\n"))
+	for i, tc := range binaryExprTestCases {
```
Comment (roman-khimov) [inline pkg/compiler/for_test.go:705]: `forLoopTestCases`

File: pkg/compiler/binary_expr_test.go [hunk 3]
```diff
@@ -267,7 +227,24 @@ var binaryExprTestCases = []testCase{
...
+	var testCases []testCase
+	srcBuilder := bytes.NewBuffer([]byte("package testcase\n"))
+	for i, tc := range binaryExprTestCases {
+		srcBuilder.WriteString(fmt.Sprintf(tc.src, i))
+	}
+
+	ne, di, err := compiler.CompileWithOptions("file.go", strings.NewReader(srcBuilder.String()), nil)
+	require.NoError(t, err)
+
+	for i, tc := range testCases {
```
Comment (roman-khimov) [inline pkg/compiler/binary_expr_test.go:239]: `binaryExprTestCases`

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/compiler/for_test.go (hunk 1 above)
Comment (fyrchik) [reply pkg/compiler/for_test.go:712]: Fixed.

File: pkg/compiler/for_test.go (hunk 2 above)
Comment (fyrchik) [reply pkg/compiler/for_test.go:705]: Fixed.

File: pkg/compiler/binary_expr_test.go (hunk 3 above)
Comment (fyrchik) [reply pkg/compiler/binary_expr_test.go]: Fixed

File: pkg/compiler/for_test.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/compiler/for_test.go:728]: Not really.

Comment (fyrchik) [reply pkg/compiler/for_test.go]: Missed the second line, fixed.

**Review by AnnaShaleva — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**